### PR TITLE
parallelization lib - for doing io bound, unbounded parallel work

### PR DESCRIPTION
**Pull request by abhishekparwal**
this library works in few modes: a) for doing short io bound work default, it spawns 2x goroutines of CPU. b) If its unsuitable parallel count can be specified. c) for running long-running io bound work provides a function to spawns unbounded goroutine.

it returns early with the first error it sees. note, first error it sees from parallel goroutine and **not** first error as in parallel many goroutines may simultaneously return an error.

### CODE REVIEW

File: parallelize/parallelize.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+					data: res,
+					err:  err,
+				}
+			}(wrk)
+		}
+	}()
+}
+
+// NewIOBoundedRunner creates bounded goroutines by factor of available CPU
+func NewIOBoundedRunner(workSize int, ioBound bool) *Runner {
```
Comment (usmyth) [inline parallelize/parallelize.go:46]: do we need an ioBound param for NewIOBoundedRunner? Seems a bit redundant

File: parallelize/parallelize.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	parallelCount := runtime.NumCPU()
+	// go routine busy doing io would be swapped out, hence 2x.
+	if ioBound {
+		parallelCount = parallelCount * 2
```
Comment (usmyth) [inline parallelize/parallelize.go:50]: benchmark tests of thriftrw, factor of 4 was found to be optimal

File: parallelize/parallelize.go (hunk 1 above)
Comment (abhishekparwal) [reply parallelize/parallelize.go]: thanks changed the naming of fnc to reflect what it does. added comments to make it obvious.

File: parallelize/parallelize.go (hunk 2 above)
Comment (abhishekparwal) [reply parallelize/parallelize.go]: cool, made it 4x for short work. for large work use a parallel count c'tor.

File: parallelize/parallelize.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+					data: res,
+					err:  err,
+				}
+			}(wrk)
+		}
+	}()
+}
+
+// NewFixedBoundedRunner creates fixed bounded goroutines by factor of available CPU based on default setting.
+// default setting is tot have parallelization as factor of 4x if iobound work which is supposed to be short.
```
Comment (smb158) [inline parallelize/parallelize.go:46]: typo, tot should be to

File: parallelize/parallelize.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+// if its  a long work use NewBoundedRunner instead with optimal config.
```
Comment (smb158) [inline parallelize/parallelize.go:47]: extra space

File: parallelize/parallelize.go (hunk 3 above)
Comment (abhishekparwal) [reply parallelize/parallelize.go:46]: since already landed will fix in next PR

File: parallelize/parallelize.go (hunk 4 above)
Comment (abhishekparwal) [reply parallelize/parallelize.go:47]: since already landed will fix in next PR

**Pull request merged**